go/src/koding/socialapi/transport.go

```diff
@@ -0,0 +1,320 @@
+		return errors.New("empty client ID")
+	}
+	return nil
+}
+
+// WithRequest attaches the session to the given request.
+func (s *Session) WithRequest(req *http.Request) *http.Request {
+	return req.WithContext(context.WithValue(req.Context(), SessionContextKey, s))
+}
+
+func (s *Session) writeTo(req *http.Request) {
+	req.Header.Set("Authorization", "Bearer "+s.ClientID)
+}
+
+func (s *Session) readFrom(req *http.Request) {
+	if auth := req.Header.Get("Authorization"); strings.HasPrefix(auth, "Bearer ") {
+		s.ClientID = auth[len("Bearer "):]
+	}
+}
+
+// AuthFunc is used to fetch session information.
+//
+// When Refresh is true, AuthFunc must obtain
+// a session directly from an authorisation
+// endpoint, in case it was cached.
+type AuthFunc func(*AuthOptions) (*Session, error)
+
+// AuthOptions represents the arguments for AuthFunc.
+type AuthOptions struct {
+	// Session represents either a valid session or
+	// Username/Team pair to request a new session
+	// from AuthFunc.
+	//
+	// If the session is valid and Refresh false,
+	// then it should be returned and AuthFunc
+	// should behave as nop.
+	//
+	// When Refresh is true, AuthFunc should request
+	// new session, no matter whether Session is
+	// valid or not.
+	Session *Session
+
+	// Refresh requests to refresh a session when true.
+	Refresh bool
+
+	// Request is a copy of the request, for which
+	// we're building authorisation.
+	Request *http.Request
+}
+
+type httpTransport interface {
+	http.RoundTripper
+	httpRequestCanceler
+	httpIdleConnectionsCloser
+}
+
+type httpRequestCanceler interface {
+	CancelRequest(*http.Request)
+}
+
+type httpIdleConnectionsCloser interface {
+	CloseIdleConnections()
+}
+
+// nopCloser is like ioutil.NopCloser, but wraps io.ReadSeeker.
+type nopCloser struct{ io.ReadSeeker }
+
+func (nopCloser) Close() (_ error) { return }
+
+var _ io.ReadSeeker = nopCloser{bytes.NewReader(nil)}
+
+// Transport is a signing transport that
+// authorizes each request with jSession.clientId.
+type Transport struct {
+	http.RoundTripper          // a transport to use; optional, by default http.DefaultClient.Transport
+	AuthFunc          AuthFunc // clientID authorisation to use; required
+	Host              string   // original Host name, overwrites req.Host; optional
+	RetryNum          int      // number of retries in case of temporary failures; optional, by default 3
+}
+
+var _ httpTransport = (*Transport)(nil)
+
+// NewSingleClient returns a HTTP transport that authenticates all
+// requests with the given session.
+//
+// Session is expected to be have a valid clientID. In order to be
+// able to refresh the session if it gets invalidated, the session
+// must also contain Username / Team information.
+//
+// This is usefull for api clients that give no control over HTTP requests,
+// like the swagger-generated ones (remote.api).
+func (t *Transport) NewSingleClient(session *Session) http.RoundTripper {
+	return &DirectorTransport{
```

Comment:
Hmm don't know the current status but this may interest you(first 40secs): https://youtu.be/FARQMJndUn0?t=1151